### PR TITLE
Fix AWS ECS

### LINKED ISSUE #33771: Segmentation fault on ECS Fargate when reading from S3
**Describe what's wrong**

I am trying to execute `select * from s3(…)` and ClickHouse dies with the following output:

```
SELECT *
FROM s3('…', 'TSV', 'root String')

Query id: 86972283-b218-493b-a56b-c1125f083ce6

[85b9a88609d0454f8fb03fa59362654d-2580379071] 2022.01.19 07:46:56.066552 [ 251 ] <Fatal> BaseDaemon: ########################################
[85b9a88609d0454f8fb03fa59362654d-2580379071] 2022.01.19 07:46:56.066602 [ 251 ] <Fatal> BaseDaemon: (version 21.12.3.32 (official build), build id: FA4A7F489F3FF6E3) (from thread 97) (query_id: 86972283-b218-493b-a56b-c1125f083ce6) Received signal Segmentation fault (11)
[85b9a88609d0454f8fb03fa59362654d-2580379071] 2022.01.19 07:46:56.066622 [ 251 ] <Fatal> BaseDaemon: Address: 0x7f2172f7a5c0 Access: read. Attempted access has violated the permissions assigned to the memory area.
[85b9a88609d0454f8fb03fa59362654d-2580379071] 2022.01.19 07:46:56.066636 [ 251 ] <Fatal> BaseDaemon: Stack trace: 0x1257420f 0x7f2207ded3c0
[85b9a88609d0454f8fb03fa59362654d-2580379071] 2022.01.19 07:46:56.066666 [ 251 ] <Fatal> BaseDaemon: 0. ? @ 0x1257420f in /usr/bin/clickhouse
[85b9a88609d0454f8fb03fa59362654d-2580379071] 2022.01.19 07:46:56.066700 [ 251 ] <Fatal> BaseDaemon: 1. ? @ 0x7f2207ded3c0 in ?
[85b9a88609d0454f8fb03fa59362654d-2580379071] 2022.01.19 07:46:56.194581 [ 251 ] <Fatal> BaseDaemon: Calculated checksum of the binary: 5BEBF5792A40F7E345921EDA3698245B. There is no information about the reference checksum.
Exception on client:
Code: 32. DB::Exception: Attempt to read after eof: while receiving packet from 10.0.1.32:9000. (ATTEMPT_TO_READ_AFTER_EOF)

Connecting to 10.0.1.32:9000 as user default.
Code: 210. DB::NetException: Connection refused (10.0.1.32:9000). (NETWORK_ERROR)
```

**Does it reproduce on recent release?**

I use Docker image `yandex/clickhouse-server:21.12.3.32`.

**How to reproduce**

I have defined the following Dockerfile:

```
FROM yandex/clickhouse-server:21.12.3.32
RUN echo '<clickhouse><s3><s3><endpoint>https://s3.eu-central-1.amazonaws.com/my-bucket/some/prefix/</endpoint><use_environment_credentials>true</use_environment_credentials></s3></s3></clickhouse>' \
        > /etc/clickhouse-server/config.d/s3.xml
```

All it does is sets `use_environment_credentials=true` for my bucket in `eu-central-1` AWS region. This is required since I am running ClickHouse on ECS Fargate (in a container) with an IAM role attached to a task.

Here is service definition in Terraform:
```terraform
locals {
  clickhouse-server-http-port   = 8123
  clickhouse-server-native-port = 9000
}

resource "aws_ecs_task_definition" "clickhouse-server" {
  family                   = "clickhouse-server"
  cpu                      = 2048
  memory                   = 8192
  network_mode             = "awsvpc"
  execution_role_arn       = aws_iam_role.ecs-exec.arn
  requires_compatibilities = ["FARGATE"]
  task_role_arn            = aws_iam_role.clickhouse-access-s3.arn  # HERE I ATTACH A ROLE TO USE TO ACCESS S3
  container_definitions    = jsonencode([
    {
      name             = "clickhouse-server"
      image            = # my custom image
      portMappings     = [
        { containerPort = local.clickhouse-server-http-port },
        { containerPort = local.clickhouse-server-native-port },
      ]
      logConfiguration = {
        logDriver = "awslogs"
        options   = {
          awslogs-group         = aws_cloudwatch_log_group.clickhouse-server.name
          awslogs-region        = data.aws_region.current.name
          awslogs-stream-prefix = "clickhouse-server"
        }
      }
    }
  ])
}

resource "aws_iam_role" "clickhouse-access-s3" {
  name               = "clickhouse-access-s3"
  assume_role_policy = data.aws_iam_policy_document.assume-ecs-execution-role.json
  inline_policy {
    name   = "clickhouse-access-s3"
    policy = data.aws_iam_policy_document.clickhouse-access-s3.json
  }
}
data "aws_iam_policy_document" "clickhouse-access-s3" {
  statement {
    effect    = "Allow"
    actions   = ["s3:*"]
    resources = []  # proper S3 permissions here
  }
}

resource "aws_ecs_cluster" "clickhouse-server" {
  name = "clickhouse-server"
}

resource "aws_ecs_service" "clickhouse-server" {
  name            = "clickhouse-server"
  launch_type     = "FARGATE"
  desired_count   = 1
  task_definition = aws_ecs_task_definition.clickhouse-server.arn
  cluster         = aws_ecs_cluster.clickhouse-server.arn
  # proper network_configuration {…}
}

resource "aws_cloudwatch_log_group" "clickhouse-server" {
  name = "clickhouse-server"
}
```

When using s3 table function without config file (running container from pure `yandex/clickhouse-server:21.12.3.32` image without my custom Dockerfile), I was able to read this file from S3 by providing explicit AWS credentials. Now, when configuration is added, the operation fails even with AWS credentials explicitly provided. The error is the same: see traceback above, ClickHouse container dies.

Solution also works when container is run using `docker run` simply on an EC2 instance (without Fargate service). No error also appears if the service is run without task role attached (`task_role_arn=null` above).

**Expected behavior**

I expect s3 table function to use `$AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` when `use_environment_credentials` enabled, and looks like codebase supports it:

https://github.com/ClickHouse/ClickHouse/blob/35883e0dae7be1ffa8948e5c56a168262fc7366f/src/IO/S3Common.cpp#L524-L526

[AWS documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) on AWS_CONTAINER_CREDENTIALS_RELATIVE_URI.

S3 endpoint settings [documentation](https://clickhouse.com/docs/en/engines/table-engines/integrations/s3/#endpoint-settings).

### PR DESCRIPTION
**Pull request by alexey-milovidov**
This was reverted

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This change was reverted

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: Normal Builds
- [ ]  Allow: Special Builds
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/IO/S3/Credentials.h [hunk 1]
```diff
@@ -19,9 +19,6 @@ namespace DB::S3
 {
 
 inline static constexpr uint64_t DEFAULT_EXPIRATION_WINDOW_SECONDS = 120;
-inline static constexpr uint64_t DEFAULT_CONNECT_TIMEOUT_MS = 1000;
```
Comment (alexey-milovidov) [inline src/IO/S3/Credentials.h:22]: Unused variables.

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 89d2de28729fc4593c80a429783019d70f4f8169 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65164/89d2de28729fc4593c80a429783019d70f4f8169/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/IO/S3/Client.cpp [hunk 2]
```diff
@@ -972,10 +972,10 @@ PocoHTTPClientConfiguration ClientFactory::createClientConfiguration( // NOLINT
 {
     auto context = Context::getGlobalContextInstance();
     chassert(context);
-    auto proxy_configuration_resolver = DB::ProxyConfigurationResolverProvider::get(DB::ProxyConfiguration::protocolFromString(protocol), context->getConfigRef());
```
Comment (alexey-milovidov) [inline src/IO/S3/Client.cpp:975]: Extraneous namespace qualification.

File: src/IO/S3/PocoHTTPClientFactory.cpp [hunk 3]
```diff
@@ -15,7 +15,10 @@ namespace DB::S3
 std::shared_ptr<Aws::Http::HttpClient>
 PocoHTTPClientFactory::CreateHttpClient(const Aws::Client::ClientConfiguration & client_configuration) const
 {
-    return std::make_shared<PocoHTTPClient>(static_cast<const PocoHTTPClientConfiguration &>(client_configuration));
```
Comment (alexey-milovidov) [inline src/IO/S3/PocoHTTPClientFactory.cpp:18]: The bug is here.

**Review by serxa — APPROVED** [review]

**Comment by tavplubix** [comment]
@Avogar, it breaks CI, please resubmit

**Comment by Avogar** [comment]
> @Avogar, it breaks CI, please resubmit

Can you share what is broken?

**Comment by Avogar** [comment]
I resolved conflicts in https://github.com/ClickHouse/clickhouse-private/pull/10963, maybe we can wait for tests there and merge it instead of revert?

**Comment by tavplubix** [comment]
> UPD: A Sync, now I see. It wasn't red and I didn't notice it

Yes, there's a bug in A Sync status that makes it yellow when it's actually red, so yellow is worth attention too, but it's not obvious. cc: @maxknv

**Comment by jakedaleweb** [comment]
Hi team. Thanks so much for addressing this, I am very keen to test it out. Please let me know when there is a container build I can use. It looks like head-alpine is currently a few commits behind this one.

**Comment by jakedaleweb** [comment]
Hello, seeing segmentation faults with this in-place still

```
18 June 2024 at 14:56 (UTC+12:00)
2024.06.18 02:56:03.241077 [ 57 ] {} <Fatal> Application: Child process was terminated by signal 11.
pollution-atomic-1-clickhouse-fargate
18 June 2024 at 14:55 (UTC+12:00)
2024.06.18 02:55:58.098447 [ 715 ] {} <Fatal> BaseDaemon: Integrity check of the executable successfully passed (checksum: C151D080117C79E61B8F56759CC284F0)
pollution-atomic-1-clickhouse-fargate
18 June 2024 at 14:55 (UTC+12:00)
2024.06.18 02:55:58.098656 [ 715 ] {} <Information> SentryWriter: Not sending crash report
pollution-atomic-1-clickhouse-fargate
18 June 2024 at 14:55 (UTC+12:00)
2024.06.18 02:55:58.098754 [ 715 ] {} <Fatal> BaseDaemon: Report this error to https://github.com/ClickHouse/ClickHouse/issues
pollution-atomic-1-clickhouse-fargate
18 June 2024 at 14:55 (UTC+12:00)
2024.06.18 02:55:57.735008 [ 715 ] {} <Fatal> BaseDaemon: (version 24.6.1.4010 (official build), build id: FF9078AF95645BE01B16663A1B04A9FF803EA617, git hash: 856f5c395a5db5a577e928000da18506f2ac0c42) (from thread 60) (no query) Received signal Segmentation fault (11)
pollution-atomic-1-clickhouse-fargate
18 June 2024 at 14:55 (UTC+12:00)
2024.06.18 02:55:57.735051 [ 715 ] {} <Fatal> BaseDaemon: Address: 0x322bcb0. Access: read. Attempted access has violated the permissions assigned to the memory area.
pollution-atomic-1-clickhouse-fargate
18 June 2024 at 14:55 (UTC+12:00)
2024.06.18 02:55:57.735096 [ 715 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000000d25d72c 0x00007f0e7274b420
pollution-atomic-1-clickhouse-fargate
18 June 2024 at 14:55 (UTC+12:00)
2024.06.18 02:55:57.735162 [ 715 ] {} <Fatal> BaseDaemon: 0. signalHandler(int, siginfo_t*, void*) @ 0x000000000d25d72c
pollution-atomic-1-clickhouse-fargate
18 June 2024 at 14:55 (UTC+12:00)
2024.06.18 02:55:57.735203 [ 715 ] {} <Fatal> BaseDaemon: 1. ? @ 0x00007f0e7274b420
```

Using the following configuration

```
<clickhouse>
    <logger>
      <level>information</level>
      <console>true</console>
    </logger>
    <storage_configuration>
      <disks>
        <s3_disk>
          <type>s3</type>
          <endpoint from_env="CLICKHOUSE_BACKEND_BUCKET"/>
          <use_environment_credentials>true</use_environment_credentials>
          <metadata_path>/var/lib/clickhouse/disks/s3_disk/</metadata_path>
        </s3_disk>
        <s3_cache>
          <type>cache</type>
          <disk>s3_disk</disk>
          <path>/var/lib/clickhouse/disks/s3_cache/</path>
          <max_size>10Gi</max_size>
        </s3_cache>
      </disks>
      <policies>
        <s3_main>
          <volumes>
            <main>
              <disk>s3_disk</disk>
            </main>
          </volumes>
        </s3_main>
      </policies>
    </storage_configuration>
</clickhouse>
```

**Comment by alexey-milovidov** [comment]
@jakedaleweb, it is not merged.

**Comment by jakedaleweb** [comment]
ah sorry! getting ahead of myself

**Comment by Sharu95** [comment]
Is this fix continued somewhere else?

**Comment by Avogar** [comment]
Here is the resubmit: https://github.com/ClickHouse/ClickHouse/pull/65362

**Comment by alexey-milovidov** [comment]
@Sharu95, we found two bugs in the AWS SDK - it is incompatible with C++23, it is not memory safe :laughing: and has data races. Pending fixes.

**Pull request merged**